sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp

```diff
@@ -121,91 +143,532 @@ struct SptrsvWrap {
+
+    using CVector = Kokkos::View<
+        const scalar_t *, Layout, typename ValuesType::device_type,
+        Kokkos::MemoryTraits<Kokkos::Unmanaged | Kokkos::RandomAccess>>;
+
+    static constexpr size_type MAX_VEC_SIZE = 11;
+    static constexpr size_type BUFF_SIZE    = 128;
+
+    using reftype = Vector;
+
+    struct ArrayType {
+      scalar_t m_data[MAX_VEC_SIZE];
+
+      KOKKOS_INLINE_FUNCTION
+      ArrayType() { init(); }
+
+      KOKKOS_INLINE_FUNCTION
+      ArrayType(const ArrayType &rhs_) {
+        for (size_type i = 0; i < MAX_VEC_SIZE; ++i) m_data[i] = rhs_.m_data[i];
+      }
+
+      KOKKOS_INLINE_FUNCTION
+      ArrayType(const Vector &) { init(); }
+
+      KOKKOS_INLINE_FUNCTION
+      void init() {
+        for (size_type i = 0; i < MAX_VEC_SIZE; ++i) m_data[i] = 0;
+      }
+
+      KOKKOS_INLINE_FUNCTION
+      ArrayType &operator+=(const ArrayType &rhs_) {
+        for (size_type i = 0; i < MAX_VEC_SIZE; ++i)
+          m_data[i] += rhs_.m_data[i];
+        return *this;
+      }
+
+      KOKKOS_INLINE_FUNCTION
+      ArrayType &operator+=(const values_t &rhs_) {
+        for (int i = 0; i < rhs_.size(); ++i) m_data[i] += rhs_(i);
+        return *this;
+      }
+    };
+
+    struct SumArray {
+      using reducer    = SumArray;
+      using value_type = ArrayType;
+      using result_view_type =
+          Kokkos::View<value_type *, execution_space, Kokkos::MemoryUnmanaged>;
+
+     private:
+      value_type &m_value;
+
+     public:
+      KOKKOS_INLINE_FUNCTION
+      SumArray(value_type &value) : m_value(value) {}
+
+      KOKKOS_INLINE_FUNCTION
+      void join(value_type &dest, const value_type &src) const { dest += src; }
+
+      KOKKOS_INLINE_FUNCTION
+      void init(value_type &val) const { val.init(); }
+
+      KOKKOS_INLINE_FUNCTION
+      value_type &reference() const { return m_value; }
+
+      KOKKOS_INLINE_FUNCTION
+      result_view_type view() const { return result_view_type(&m_value, 1); }
+
+      KOKKOS_INLINE_FUNCTION
+      bool reference_scalar() const { return true; }
+    };
+
+    RowMapType row_map;
+    EntriesType entries;
+    ValuesType values;
+    LHSType lhs;
+    RHSType rhs;
+    entries_t nodes_grouped_by_level;
+    size_type block_size;
+    size_type block_items;
+
+    Common(const RowMapType &row_map_, const EntriesType &entries_,
+           const ValuesType &values_, LHSType &lhs_, const RHSType &rhs_,
+           const entries_t &nodes_grouped_by_level_,
+           const size_type block_size_)
+        : row_map(row_map_),
+          entries(entries_),
+          values(values_),
+          lhs(lhs_),
+          rhs(rhs_),
+          nodes_grouped_by_level(nodes_grouped_by_level_),
+          block_size(block_size_),
+          block_items(block_size * block_size) {
+      KK_REQUIRE_MSG(block_size > 0,
+                     "Tried to use block_size=0 with the blocked Common?");
+      KK_REQUIRE_MSG(block_size <= MAX_VEC_SIZE,
```

Comment:
I am not sure that our codes will actually verify this condition?
Are you simply looping over chunks of size 11 within the blocks?

Reply:
Same thoughts as above, I can increase BUFF_SIZE to 256 or even 512 if we want. I don't think either choice will impact performance for smaller blocks. That would allow us to go up to block size 16 or 22.